File: cpp/examples/interop_tool.cpp
// 跨语言互操作工具：与 Python 参考客户端交叉验证协议字节。
//
// 用法：
//   rmq_interop --emit                 打印规范命令的 JSON / ROCKETMQ 两种编码（十六进制）
//   rmq_interop --decode <hex>         解码给定十六进制帧并打印字段（供 Python -> C++ 验证）
//   rmq_interop --emit-route           打印规范 TopicRouteData 的 JSON
//   rmq_interop --emit-heartbeat       打印规范 HeartbeatData 的 JSON
//   rmq_interop --decode-route <hex>   按 TopicRouteData 解码 hex(body)，回吐重序列化 JSON
//   rmq_interop --decode-heartbeat <hex>  按 HeartbeatData 解码 hex(body)，回吐重序列化 JSON
//
// "解码后重新序列化再交给 Python 用 json.loads 比对"这种形式，
// 能把"字段名不一致"暴露成缺失键，而不是靠字符串相等这种脆弱断言。
//
// 规范命令（两侧必须一致）：
//   code=10, language=1(CPP), opaque=7, version=0, flag=0,
//   remark="hello", extFields{topic=T_TEST, queueId=2}, body="payload"
#include <iostream>
#include <map>
#include <string>

#include "rocketmq/common/subscription_data.h"
#include "rocketmq/common/sysflag.h"
#include "rocketmq/common/util_all.h"
#include "rocketmq/remoting/protocol/heartbeat.h"
#include "rocketmq/remoting/protocol/remoting_command.h"
#include "rocketmq/remoting/protocol/route.h"

using namespace rocketmq;

namespace {

RemotingCommand canonicalCommand(uint8_t serializeType) {
    RemotingCommand cmd;
    cmd.code = 10;
    cmd.language = LanguageCode::CPP;
    cmd.version = 0;
    cmd.opaque = 7;
    cmd.flag = 0;
    cmd.remark = "hello";
    cmd.hasRemark = true;
    cmd.addExtField("topic", "T_TEST");
    cmd.addExtField("queueId", "2");
    cmd.body = "payload";
    cmd.hasBody = true;
    cmd.serializeTypeCurrentRPC = serializeType;
    return cmd;
}

TopicRouteData canonicalRoute() {
    TopicRouteData t;
    t.setOrderTopicConf("");
    t.queueDatas.push_back(
        QueueData("broker-a", 4, 4, PermName::PERM_READ | PermName::PERM_WRITE, 0));
    t.queueDatas.push_back(QueueData("broker-b", 2, 2, PermName::PERM_READ, 0));
    std::map<int64_t, std::string> addrsA;
    addrsA[0] = "127.0.0.1:10911";
    addrsA[1] = "127.0.0.1:10912";
    std::map<int64_t, std::string> addrsB;
    addrsB[0] = "127.0.0.1:10913";
    t.brokerDatas.push_back(BrokerData("DefaultCluster", "broker-a", addrsA, "", false));
    t.brokerDatas.push_back(BrokerData("DefaultCluster", "broker-b", addrsB, "", false));
    return t;
}

HeartbeatData canonicalHeartbeat() {
    HeartbeatData hb("10.0.0.1@12345");
    hb.addProducerData(ProducerData("pg_probe"));
    ConsumerData cd("cg_probe", ConsumeType::CONSUME_PASSIVELY, MessageModel::CLUSTERING,
                    ConsumeFromWhere::CONSUME_FROM_LAST_OFFSET);
    cd.setUnitMode(false);
    SubscriptionData sd = FilterAPI::buildSubscriptionData("TopicProbe", "TagA||TagB");
    sd.setSubVersion(1700000000000LL);
    cd.addSubscriptionData(sd);
    hb.addConsumerData(cd);
    return hb;
}

// 解码 hex 后按指定结构重序列化，输出单行 JSON
template <typename DecodeFn>
int decodeAndDump(const char* hex, DecodeFn fn) {
    Bytes raw = UtilAll::string2Bytes(hex);
    if (raw.empty()) {
        std::cerr << "empty/invalid hex\n";
        return 2;
    }
    JsonValue out;
    if (!fn(raw, out)) {
        std::cerr << "decode failed\n";
        return 3;
    }
    std::cout << out.dump() << "\n";
    return 0;
}

}  // namespace

int main(int argc, char** argv) {
    if (argc < 2) {
        std::cerr << "usage: rmq_interop --emit | --decode <hex> | --emit-route | "
                     "--emit-heartbeat | --decode-route <hex> | --decode-heartbeat <hex>\n";
        return 1;
    }
    const std::string mode = argv[1];

    if (mode == "--emit") {
        RemotingCommand jsonCmd = canonicalCommand(SerializeType::JSON);
        RemotingCommand rmqCmd = canonicalCommand(SerializeType::ROCKETMQ);
        std::cout << "JSON " << UtilAll::bytes2String(jsonCmd.encode()) << "\n";
        std::cout << "ROCKETMQ " << UtilAll::bytes2String(rmqCmd.encode()) << "\n";
        return 0;
    }

    if (mode == "--emit-route") {
        std::cout << canonicalRoute().toJson().dump() << "\n";
        return 0;
    }

    if (mode == "--emit-heartbeat") {
        std::cout << canonicalHeartbeat().toJson().dump() << "\n";
        return 0;
    }

    if (mode == "--decode" && argc >= 3) {
        Bytes raw = UtilAll::string2Bytes(argv[2]);
        if (raw.empty()) {
            std::cerr << "empty/invalid hex\n";
            return 2;
        }
        RemotingCommand cmd;
        std::string err;
        if (!RemotingCommand::tryDecode(raw, cmd, &err)) {
            std::cerr << "decode failed: " << err << "\n";
            return 3;
        }
        std::cout << "code=" << cmd.code
                  << ";language=" << static_cast<int>(cmd.language)
                  << ";opaque=" << cmd.opaque
                  << ";version=" << cmd.version
                  << ";remark=" << cmd.remark
                  << ";body=" << cmd.body;
        for (const auto& kv : cmd.extFields) {
            std::cout << ";ext." << kv.first << "=" << kv.second;
        }
        std::cout << "\n";
        return 0;
    }

    if (mode == "--decode-route" && argc >= 3) {
        return decodeAndDump(argv[2], [](const Bytes& raw, JsonValue& out) {
            TopicRouteData t;
            if (!TopicRouteData::decode(raw, t)) {
                return false;
            }
            out = t.toJson();
            return true;
        });
    }

    if (mode == "--decode-heartbeat" && argc >= 3) {
        return decodeAndDump(argv[2], [](const Bytes& raw, JsonValue& out) {
            HeartbeatData hb;
            if (!HeartbeatData::decode(raw, hb)) {
                return false;
            }
            out = hb.toJson();
            return true;
        });
    }

    std::cerr << "unknown mode: " << mode << "\n";
    return 1;
}

File: dotnet/examples/RocketMQ.Examples/InteropTool.cs
// 跨语言互操作工具：与 Python 参考客户端交叉验证协议字节。
// 对应 cpp/examples/interop_tool.cpp（172 行）。
//
// 用法（由 Program 以 "interop <mode> [args...]" 形式调用）：
//   rmq interop --emit                 打印规范命令的 JSON / ROCKETMQ 两种编码（十六进制）
//   rmq interop --decode <hex>         解码给定十六进制帧并打印字段（供 Python -> C# 验证）
//   rmq interop --emit-route           打印规范 TopicRouteData 的 JSON
//   rmq interop --emit-heartbeat       打印规范 HeartbeatData 的 JSON
//   rmq interop --decode-route <hex>   按 TopicRouteData 解码 hex(body)，回吐重序列化 JSON
//   rmq interop --decode-heartbeat <hex>  按 HeartbeatData 解码 hex(body)，回吐重序列化 JSON
//
// "解码后重新序列化再交给 Python 用 json.loads 比对"这种形式，
// 能把"字段名不一致"暴露成缺失键，而不是靠字符串相等这种脆弱断言。
using System;
using System.Text;
using RocketMQ.Client;
using RocketMQ.Common;
using RocketMQ.Remoting.Protocol;

namespace RocketMQ.Examples;

/// <summary>跨语言协议字节互操作（与 Python 参考客户端对齐）。</summary>
internal static class InteropTool
{
    // 规范命令（两侧必须一致）：
    //   code=10, language=1(CPP), opaque=7, version=0, flag=0,
    //   remark="hello", extFields{topic=T_TEST, queueId=2}, body="payload"
    private const int CanonicalCode = 10;

    public static int Run(string[] args)
    {
        if (args.Length < 1)
        {
            Console.Error.WriteLine("usage: rmq interop --emit | --decode <hex> | --emit-route | " +
                                    "--emit-heartbeat | --decode-route <hex> | --decode-heartbeat <hex>");
            return 1;
        }

        string mode = args[0];
        switch (mode)
        {
            case "--emit":
                return Emit();
            case "--emit-route":
                Console.WriteLine(Json.Dump(CanonicalRoute().ToJson()));
                return 0;
            case "--emit-heartbeat":
                Console.WriteLine(Json.Dump(CanonicalHeartbeat().ToJson()));
                return 0;
            case "--decode" when args.Length >= 2:
                return DecodeFrame(args[1]);
            case "--decode-route" when args.Length >= 2:
                return DecodeRoute(args[1]);
            case "--decode-heartbeat" when args.Length >= 2:
                return DecodeHeartbeat(args[1]);
            default:
                Console.Error.WriteLine("unknown mode: " + mode);
                return 1;
        }
    }

    // ---------------- --emit ----------------

    private static int Emit()
    {
        RemotingCommand jsonCmd = BuildCanonicalCommand(SerializeType.Json);
        RemotingCommand rmqCmd = BuildCanonicalCommand(SerializeType.Rocketmq);
        Console.WriteLine("JSON " + UtilAll.Bytes2String(jsonCmd.Encode()));
        Console.WriteLine("ROCKETMQ " + UtilAll.Bytes2String(rmqCmd.Encode()));
        return 0;
    }

    private static RemotingCommand BuildCanonicalCommand(byte serializeType)
    {
        var cmd = new RemotingCommand
        {
            Code = CanonicalCode,
            Language = LanguageCode.Cpp,
            Version = 0,
            Opaque = 7,
            Flag = 0,
            Remark = "hello",
            HasRemark = true,
            SerializeTypeCurrentRpc = serializeType,
        };
        cmd.AddExtField("topic", "T_TEST");
        cmd.AddExtField("queueId", "2");
        cmd.Body = Encoding.UTF8.GetBytes("payload");
        cmd.HasBody = true;
        return cmd;
    }

    // ---------------- --emit-* 规范数据 ----------------

    private static TopicRouteData CanonicalRoute()
    {
        var route = new TopicRouteData { OrderTopicConf = string.Empty };
        route.QueueDatas.Add(new QueueData("broker-a", 4, 4, PermName.PermRead | PermName.PermWrite, 0));
        route.QueueDatas.Add(new QueueData("broker-b", 2, 2, PermName.PermRead, 0));
        var addrsA = new SortedDictionary<long, string>
        {
            [0] = "127.0.0.1:10911",
            [1] = "127.0.0.1:10912",
        };
        var addrsB = new SortedDictionary<long, string>
        {
            [0] = "127.0.0.1:10913",
        };
        route.BrokerDatas.Add(new BrokerData("DefaultCluster", "broker-a", addrsA, string.Empty, false));
        route.BrokerDatas.Add(new BrokerData("DefaultCluster", "broker-b", addrsB, string.Empty, false));
        return route;
    }

    private static HeartbeatData CanonicalHeartbeat()
    {
        var hb = new HeartbeatData("10.0.0.1@12345");
        hb.AddProducerData(new ProducerData("pg_probe"));
        var cd = new ConsumerData("cg_probe", ConsumeType.ConsumePassively, MessageModel.Clustering,
            ConsumeFromWhere.ConsumeFromLastOffset)
        {
            UnitMode = false,
        };
        SubscriptionData sd = FilterAPI.BuildSubscriptionData("TopicProbe", "TagA||TagB");
        sd.SubVersion = 1700000000000L;
        cd.AddSubscriptionData(sd);
        hb.AddConsumerData(cd);
        return hb;
    }

    // ---------------- --decode ----------------

    private static int DecodeFrame(string hex)
    {
        byte[] raw = UtilAll.String2Bytes(hex);
        if (raw.Length == 0)
        {
            Console.Error.WriteLine("empty/invalid hex");
            return 2;
        }

        RemotingCommand cmd;
        try
        {
            cmd = RemotingCommand.Decode(raw);
        }
        catch (Exception e)
        {
            Console.Error.WriteLine("decode failed: " + e.Message);
            return 3;
        }

        var sb = new StringBuilder();
        sb.Append("code=").Append(cmd.Code.ToString(CultureInfo.InvariantCulture));
        sb.Append(";language=").Append(cmd.Language.ToString(CultureInfo.InvariantCulture));
        sb.Append(";opaque=").Append(cmd.Opaque.ToString(CultureInfo.InvariantCulture));
        sb.Append(";version=").Append(cmd.Version.ToString(CultureInfo.InvariantCulture));
        sb.Append(";remark=").Append(cmd.Remark);
        sb.Append(";body=").Append(cmd.HasBody ? Encoding.UTF8.GetString(cmd.Body) : string.Empty);
        foreach (var kv in cmd.ExtFields)
        {
            sb.Append(";ext.").Append(kv.Key).Append('=').Append(kv.Value);
        }

        Console.WriteLine(sb.ToString());
        return 0;
    }

    private static int DecodeRoute(string hex)
    {
        byte[] raw = UtilAll.String2Bytes(hex);
        if (raw.Length == 0)
        {
            Console.Error.WriteLine("empty/invalid hex");
            return 2;
        }

        if (!TopicRouteData.Decode(raw, out TopicRouteData route))
        {
            Console.Error.WriteLine("decode failed");
            return 3;
        }

        Console.WriteLine(Json.Dump(route.ToJson()));
        return 0;
    }

    private static int DecodeHeartbeat(string hex)
    {
        byte[] raw = UtilAll.String2Bytes(hex);
        if (raw.Length == 0)
        {
            Console.Error.WriteLine("empty/invalid hex");
            return 2;
        }

        if (!HeartbeatData.Decode(raw, out HeartbeatData hb))
        {
            Console.Error.WriteLine("decode failed");
            return 3;
        }

        Console.WriteLine(Json.Dump(hb.ToJson()));
        return 0;
    }
}
